File: services/auctions/handlers/delete_lot.py
''' The `import` statement in Python is used to import modules or packages into your code.'''
import os
import json
import pymongo
import boto3
from botocore.exceptions import ClientError
headers = {
    'Content-Type': 'application/json',
    'Access-Control-Allow-Origin': '*',
    'Access-Control-Allow-Credentials': True,
    'Access-Control-Allow-Headers': '*',
    'Access-Control-Allow-Methods': '*'
}

client = pymongo.MongoClient(os.environ['MONGO_CLIENT'])
db = client[os.environ['DATABASE']]
collection = db[os.environ["LOT_COLLECTION_NAME"]]
auction_collection = db[os.environ["AUCTION_MONGODB_COLLECTION_NAME"]]
counter_collection = db[os.environ["COUNTER_LOT"]]


def extract_s3_key_from_image(image_obj):
    """
    Extract S3 key from image object
    :param image_obj: The image object with url and featured properties
    :return: The validated S3 key or None if invalid
    """
    if not image_obj or not isinstance(image_obj, dict) or 'url' not in image_obj:
        return None

    url = image_obj.get('url')
    if not url or not isinstance(url, str) or url.strip() == '':
        return None

    return url.strip()


def delete_images_from_s3(images):
    """
    Delete images from S3
    :param images: Array of image objects with url and featured properties
    :return: Success status (True if all deleted successfully, False otherwise)
    """
    if not images or not isinstance(images, list) or len(images) == 0:
        return True  # No images to delete

    try:
        # Initialize S3 client
        s3_client = boto3.client('s3', region_name=os.environ.get('REGION', 'eu-west-2'))

        bucket_name = os.environ.get('S3_BUCKET')

        # Extract S3 keys from image objects
        s3_keys = []
        for image in images:
            key = extract_s3_key_from_image(image)
            if key:
                s3_keys.append(key)

        if len(s3_keys) == 0:
            print('No valid S3 keys found in images array')
            return True

        # Delete objects from S3
        success_count = 0
        for key in s3_keys:
            try:
                s3_client.delete_object(Bucket=bucket_name, Key=f"public/{key}")
                print(f'Successfully deleted S3 object: {key}')
                success_count += 1
            except ClientError as e:
                print(f'Failed to delete S3 object {key}: {str(e)}')

        print(f'Deleted {success_count}/{len(s3_keys)} images from S3')
        return success_count == len(s3_keys)

    except Exception as e:
        print(f'Error deleting images from S3: {str(e)}')
        return False


def update_lot_numbers(auction_id, seller_email, deleted_lot_number):
    """
    Updates lot numbers after a lot is deleted to maintain sequential ordering
    """
    try:
        # Find all lots with lot number greater than deleted lot, sorted by lot number
        lots_to_update = collection.find({
            "auction_id": auction_id,
            "seller_email": seller_email,
            "lot_number": {"$gt": deleted_lot_number}
        }).sort("lot_number", 1)

        # Setting the condition for the Bulk update operation
        bulk_ops = []
        for lot in lots_to_update:
            bulk_ops.append(
                pymongo.UpdateOne(
                    {"_id": lot["_id"]},
                    {"$set": {"lot_number": lot["lot_number"] - 1}}
                )
            )

        # bulk operation to update the lot number when a lot is deleted
        if bulk_ops:
            collection.bulk_write(bulk_ops)
            return True
        return False

    except Exception as e:
        print(f"Error updating lot numbers: {str(e)}")
        return False




def delete_lot(event, context):
    """
    The function "delete_lot" is used to delete a lot.

    :param event: The event parameter is an object that contains information about the triggering event
    that caused the function to be invoked. This can include details such as the event type, event
    source, and any event-specific data
    :param context: The `context` parameter is an object that provides information about the runtime
    environment of the function. It includes details such as the function's execution ID, the function's
    name, and the function's memory limit
    """
    try:
        try:
            seller_email = event['requestContext']['authorizer']['claims']['email']
            if "cognito:groups" in event['requestContext']['authorizer']['claims'] and not 'seller' in event['requestContext']['authorizer']['claims']["cognito:groups"]:
                return {
                "statusCode": 403,
                "headers": headers,
                "body": json.dumps({"message": "You do not have access to perform this API action"})
            }
            print('email', seller_email)
        except:
            return {
                "statusCode": 403,
                "headers": headers,
                "body": json.dumps({"message": "You do not have access to perform this API action"})
            }
        # Parse the incoming JSON request
        request_body = json.loads(event['body'])
        print(request_body)

        # Check if the request includes the necessary data for lot identification
        lot_number = request_body.get('lot_number', 0)
        auction_id = request_body.get('auction_id','')

        if not lot_number or not seller_email:
            return {
                "statusCode": 400,
                'headers': headers,
                "body": json.dumps({
                    "message": "Both lot_number and seller_email are required for lot deletion."
                })
            }

        # Find the lot first to get image data before deletion
        lot = collection.find_one({
            "lot_number": lot_number, 
            "seller_email": seller_email, 
            "auction_id": auction_id
        })

        if not lot:
            return {
                "statusCode": 404,
                'headers': headers,
                "body": json.dumps({
                    "message": "Lot not found"
                })
            }

        # Delete images from S3 before deleting the lot
        if lot.get('images') and len(lot['images']) > 0:
            print(f"Deleting {len(lot['images'])} images from S3 for lot {lot_number}")
            s3_delete_success = delete_images_from_s3(lot['images'])

            if not s3_delete_success:
                print('Some images failed to delete from S3, but continuing with lot deletion')

        # Delete the specified lot from the MongoDB collection
        delete_result = collection.delete_one({
            "lot_number": lot_number, 
            "seller_email": seller_email, 
            "auction_id": auction_id
        })

        if delete_result.deleted_count == 1:
            auction_record = auction_collection.find_one({"auction_id": auction_id, "seller_email": seller_email})
            # print('auction', auction_record.get('status'))
            # Update lot numbers after deletion
            if auction_record.get('status') == 'Draft':

                # updating the lot number to maintain the order
                # the deleted lots number will be assigned to the lot which is in front of it , example lot 1 is deleted, so lot2 will be converted to lot1
                if update_lot_numbers(auction_id, seller_email, lot_number):
                    print("Lot numbers updated successfully.")

            counter_update = counter_collection.update_one(
                {"auction_id": auction_id, "seller_email": seller_email},
                {"$inc": {"starting_sequence": -1}}
            )
            if counter_update.modified_count == 1:
                print("Counter updated successfully.")



            if auction_record and "total_lots" in auction_record and auction_record["total_lots"] > 0:
                # Decrease the existing "total_lots" count
                auction_collection.update_one(
                    {"auction_id": auction_id, "seller_email": seller_email},
                    {"$inc": {"total_lots": -1}}
                )
            else:
                # Calculate the total lots count (if not already calculated) and update the auction record
                total_lots_count = collection.count_documents({"seller_email": seller_email, "auction_id": auction_id})
                auction_collection.update_one(
                    {"auction_id": auction_id, "seller_email": seller_email},
                    {"$set": {"total_lots": total_lots_count}}
                )
            return {
                "statusCode": 200,
                'headers': headers,
                "body": json.dumps({"message": " deleted successfully."})
            }
        else:
            return {
                "statusCode": 404,
                'headers': headers,
                "body": json.dumps({
                    "message": "email not found"
                })
            }
    except Exception as e:
        print('Error:', str(e))
        return {
            "statusCode": 500,
            'headers': headers,
            "body": json.dumps({"error": str(e)})
        }

File: services/in-person-auction/handlers/delete_lot.js
/* eslint-disable no-underscore-dangle */
/* eslint-disable camelcase */
/* eslint-disable no-console */
/* eslint-disable import/extensions */
/* eslint-disable import/no-unresolved */
const { ObjectId } = require('mongodb')
const AWS = require('aws-sdk')
const helpers = require('../lib/helper')
const mongoConnection = require('../lib/mongodb_helper')
const Lot = require('../entities/Lot')
const Counter = require('../entities/Counter')
const Auction = require('../entities/Auction')

let connection = null

/**
 * Extract S3 key from image object
 * @param {Object} imageObj - The image object with url and featured properties
 * @returns {string|null} - The validated S3 key or null if invalid
 */
function extractS3KeyFromImage(imageObj) {
    if (!imageObj || typeof imageObj !== 'object' || !imageObj.url) {
        return null
    }

    const { url } = imageObj
    if (typeof url !== 'string' || url.trim() === '') {
        return null
    }

    return url.trim()
}

/**
 * Delete images from S3
 * @param {Array} images - Array of image objects with url and featured properties
 * @returns {Promise<boolean>} - Success status
 */
async function deleteImagesFromS3(images) {
    if (!images || !Array.isArray(images) || images.length === 0) {
        return true // No images to delete
    }

    try {
        const s3Client = new AWS.S3({
            region: process.env.AWS_REGION || 'eu-west-2',
        })

        const bucketName = process.env.S3_BUCKET

        // Extract S3 keys from image objects
        const s3Keys = images
            .map(extractS3KeyFromImage)
            .filter((key) => key !== null)

        if (s3Keys.length === 0) {
            console.log('No valid S3 keys found in images array')
            return true
        }

        // Delete objects from S3
        const deletePromises = s3Keys.map(async (key) => {
            try {
                await s3Client.deleteObject({
                    Bucket: bucketName,
                    Key: `public/${key}`,
                }).promise()
                console.log(`Successfully deleted S3 object: ${key}`)
                return true
            } catch (error) {
                console.error(`Failed to delete S3 object ${key}:`, error)
                return false
            }
        })

        const results = await Promise.all(deletePromises)
        const successCount = results.filter((result) => result === true).length

        console.log(`Deleted ${successCount}/${s3Keys.length} images from S3`)
        return successCount === s3Keys.length
    } catch (error) {
        console.error('Error deleting images from S3:', error)
        return false
    }
}

async function updateLotNumbers(auctionId, sellerEmail, deletedLotNumber) {
    /**
   * Updates lot numbers after a lot is deleted to maintain sequential ordering
   */
    try {
        // Find all lots with lot_number greater than deleted lot
        const lotsToUpdate = await Lot.find({
            auction_id: auctionId,
            seller_email: sellerEmail,
            lot_number: { $gt: deletedLotNumber },
        }).sort({ lot_number: 1 })

        // Prepare bulk operations
        if (!lotsToUpdate || lotsToUpdate.length === 0) {
            return true
        }

        const bulkOps = lotsToUpdate.map((lot) => ({
            updateOne: {
                filter: { _id: new ObjectId(lot._id) },
                update: { $set: { lot_number: lot.lot_number - 1 } },
            },
        }))

        // Execute bulk write if there are updates
        if (bulkOps.length > 0) {
            await Lot.bulkWrite(bulkOps)
            return true
        }

        return false
    } catch (err) {
        console.error(`Error updating lot numbers: ${err.message}`)
        return false
    }
}

module.exports.delete_lot = async (event) => {
    try {
        // --- Authorization Check ---
        const { claims } = event.requestContext?.authorizer || {}
        if (!claims || !claims['cognito:username']) {
            return {
                statusCode: 403,
                headers: await helpers.getHeaders(),
                body: JSON.stringify({ message: 'You do not have access to perform this API action' }),
            }
        }
        // --- Ensure MongoDB connection ---
        if (connection === null || !connection.readyState) {
            connection = await mongoConnection.connect()
        }

        const { lot_id, auction_id } = JSON.parse(event.body) || {}

        // --- Validate request body ---
        if (!lot_id || !auction_id) {
            return {
                statusCode: 400,
                headers: await helpers.getHeaders(),
                body: JSON.stringify({ message: 'Lot ID and auction ID is required' }),
            }
        }

        // Get seller email from Cognito claims
        const email = event.requestContext.authorizer.claims['cognito:username']
        const auctionDetails = await Auction.findOne({ auction_id, seller_email: email })

        // --- Validate auction and state ---
        if (!auctionDetails) {
            return {
                statusCode: 404,
                headers: helpers.getHeaders(),
                body: JSON.stringify({ message: 'Auction not found' }),
            }
        }
        if (auctionDetails.status !== 'Draft') {
            return {
                statusCode: 400,
                headers: helpers.getHeaders(),
                body: JSON.stringify({ message: 'Auction not in draft state' }),
            }
        }

        // --- Find the lot first to get image data ---
        const lot = await Lot.findOne({ _id: new ObjectId(lot_id) })

        if (!lot) {
            return {
                statusCode: 404,
                headers: await helpers.getHeaders(),
                body: JSON.stringify({ message: 'Lot not found' }),
            }
        }

        // --- Delete images from S3 before deleting the lot ---
        if (lot.images && lot.images.length > 0) {
            console.log(`Deleting ${lot.images.length} images from S3 for lot ${lot_id}`)
            const s3DeleteSuccess = await deleteImagesFromS3(lot.images)

            if (!s3DeleteSuccess) {
                console.warn('Some images failed to delete from S3, but continuing with lot deletion')
            }
        }

        // --- Delete the lot from database ---
        await Lot.findOneAndDelete({ _id: new ObjectId(lot_id) })

        // --- Update lot counter (decrement sequence) ---
        await Counter.findOneAndUpdate({
            seller_email: email, auction_id, record_type: 'Lots',
        }, { $inc: { starting_sequence: -1 } })

        // --- Reorder remaining lots sequentially ---
        if (!(await updateLotNumbers(auction_id, email, lot.lot_number))) {
            return {
                statusCode: 400,
                headers: await helpers.getHeaders(),
                body: JSON.stringify({ message: 'Error updating lot numbers' }),
            }
        }

        // Return 204 No Content on successful deletion
        return {
            statusCode: 204,
            headers: await helpers.getHeaders(),
        }
    } catch (error) {
        console.log('Internal Server Error', error)
        return {
            headers: await helpers.getHeaders(),
            statusCode: 500,
            body: JSON.stringify({
                message: 'Internal Server Error',
            }),
        }
    }
}
